### ISSUE TITLE
Tracking issue for `box_patterns` feature

### ISSUE THREAD

**Issue by aturon** [rust-lang/rust#29641]
*[View all comments](https://triagebot.infra.rust-lang.org/gh-comments/rust-lang/rust/issues/29641)*

Box patterns were gated by [RFC 469](https://github.com/rust-lang/rfcs/pull/469).

**Comment by pyfisch** [comment]
Will it be possible to use box patterns with smart pointers?

**Comment by clarcharr** [comment]
Odd question: I've seen a few `fn method(self: Box<Self>)` type things in libstd; is there any plan to eventually turn this into `fn method(box self)` ?

**Comment by pyfisch** [comment]
Types will be written `Box<Foo>` and not `box Foo` even if box_patterns is stabilized. Allowing `box Foo` would be a new feature that is not planned I think. `box` is useful to box values (shorter then `Box::new(value)` and allows some pattern matching impossible before.

**Comment by clarcharr** [comment]
Oh, I understand that in general, `box Foo` won't be a type. I was just wondering because I was under the assumption that `&self` in method definitions was more of a pattern than a type description, which means that `box self` in methods makes sense.

**Comment by SimonSapin** [comment]
`&self` in methods does not work like a pattern. For example in `fn foo(x: &u32) { let &y = x }` the `&y` pattern "removes" one pointer indirection and make `y` have type `u32`, whereas in a `impl SomTrait for u32 { fn bar(&self) { … }}` method `self` has type `&u32` with the `&_` indirection included.

**Comment by jadbox** [comment]
Is Box patterns enabled, without flags, on Nightly?

**Comment by SimonSapin** [comment]
@jadbox No. Unless someone missed something, this issue would be closed if they were.

rustc 1.18.0-nightly (2b4c91158 2017-04-25)

```rust
error: box pattern syntax is experimental (see issue #29641)
 --> a.rs:2:9
  |
2 |     let box a = Box::new(1);
  |         ^^^^^
  |
  = help: add #![feature(box_patterns)] to the crate attributes to enable

error: aborting due to previous error
```

**Comment by jadbox** [comment]
thanks @SimonSapin 

What's left to do on the feature before it can be considered 'stable' enough for nightly?

**Comment by steveklabnik** [comment]
> What's left to do on the feature before it can be considered 'stable' enough for nightly?

This is related to box syntax more generally, so the questions around that need to be resolved first.

**Comment by deontologician** [comment]
This is really critical to matching recursive enums. I have some code like:

```rust
match act {
    Subcondition(box Always(act)) => simplify(act),
    Subcondition(box cond) => Subcondition(Box::new(simplify(cond))),
    otherwise => otherwise
}
```
And basically, I'm stuck on nightly.

Is this box pattern syntax *really* super tied to the placement syntax in the other issues that are also using the `box` keyword? The resemblance seems cosmetic. Boxing recursive enums is a basic thing to do, and being able to break those recursive enums up in match blocks seems like a pure win, regardless of how placement works.

Maybe changing the keyword from `box` would help get this feature going again?

**Comment by glaebhoerl** [comment]
As [discussed](https://github.com/rust-lang/rfcs/pull/2005#issuecomment-307935415) on the RFC for default binding modes in `match` (tracking issue https://github.com/rust-lang/rust/issues/42640), I think the best solution here would not be `box` patterns, but simply to have the `ref` "default binding mode" automatically propagate through all types with `Deref` implementations (not just through `&` itself), and likewise the `ref mut` mode through all `DerefMut` types (and not just `&mut`).

So you could do things like:

```rust
let foo = Rc::new(Some(true)); // or Box, or Arc, or ...
match &foo {
    Some(b) => { /* b is of type `&bool` here */ }
}
```

There may or may not be complications around things like `DerefPure` here, I can't remember. (If the complications don't implicate safety then I think we should just accept them with, "don't write surprising `Deref` implementations or you'll get surprising results".)

**Comment by rpjohnst** [comment]
Would it also be possible for `&` in patterns to go through `Deref`, to preserve the ability to write the "default binding mode" stuff explicitly? Or would it be better to keep `box` as the explicit version, and let it be stabilized on its own?

**Comment by glaebhoerl** [comment]
I would prefer to deprecate `&` and `&mut` (and `ref` and `ref mut`) in pattern contexts and just do this kind of adjustment on the RHS when necessary (in expression context) - so in this case, just use `*` to deref as normal - which may not always be as ideal as possible from a code-golf perspective, but is generally much clearer and less of a learning hazard.

**Comment by rpjohnst** [comment]
I don't think binding modes are as expressive on their own, though. We would need something beyond the RFC in order to fully remove `&`/`&mut`, no?

Or did you mean forcing people to add `*` in the match arm after the pattern? I feel like that's a step backwards and we don't have a clear path to avoiding that either.

**Comment by glaebhoerl** [comment]
There was some discussion about these things on the RFC, but I don't really remember it. In either case, this is getting into the weeds; I think the part that's important for now is just that the use case that was going to be satisfied by `box` patterns should be satisfied by extending default binding modes instead. After that I would *like to* also be able to remove all the other things from patterns, *if* there aren't any major obstacles (e.g. strictly reduced expressivity), but we can attend to those details later in the process (e.g. I'm imagining the first thing would end up as one RFC / feature flag and the second as another, later on, or something).

**Comment by deontologician** [comment]
@glaebhoerl So the best place to discuss this is default binding modes then? Is there much momentum there?

**Comment by glaebhoerl** [comment]
@deontologician Yes I'd think so (go ahead and drop a line there if the use case is important to you). Default binding modes itself (based on the tracking issue) is implemented but has a ways to go before stabilization; there hasn't been any "official" discussion about doing the thing I described, but a few people have seemed to agree that it seems like a promising direction to move in.

**Comment by deontologician** [comment]
Yeah, looks like discussion of that part is picked up in https://github.com/rust-lang/rfcs/issues/2099 which doesn't have an RFC yet. I'll move my focus there, thanks!

**Comment by jonhoo** [comment]
@pnkfelix Is there a planned path to stabilization for this? The feature gate landed in https://github.com/rust-lang/rust/pull/22175 quite some time ago, and this is pretty essential for doing nested matches over struct and enums that contains boxes.

**Comment by jcotton42** [comment]
Bumping this since it's been 11 months and I'm also interested in this (don't have any specific use case, just interested)

**Comment by Robbepop** [comment]
> > As [discussed](https://github.com/rust-lang/rfcs/pull/2005#issuecomment-307935415) on the RFC for default binding modes in `match` (tracking issue #42640), I think the best solution here would not be `box` patterns, but simply to have the `ref` "default binding mode" automatically propagate through all types with `Deref` implementations (not just through `&` itself), and likewise the `ref mut` mode through all `DerefMut` types (and not just `&mut`).
> > So you could do things like:
> > ```rust
> > let foo = Rc::new(Some(true)); // or Box, or Arc, or ...
> > match &foo {
> >     Some(b) => { /* b is of type `&bool` here */ }
> >     None => { ... }
> > }
> > ```
> > 
> > 
> > There may or may not be complications around things like `DerefPure` here, I can't remember. (If the complications don't implicate safety then I think we should just accept them with, "don't write surprising `Deref` implementations or you'll get surprising results".)
> 
> "Default binding mode" sounds incredibly good! Is there anything related to this that I can take a look at?
> I was looking forward to box patterns, but now you've provided an alternative that sounds even better.

I use the `box_patterns` feature quite extensively in here: https://github.com/Robbepop/stevia/blob/master/simplifier/src/simplifications/term_const_prop.rs#L37
Especially conditional move out of a structure guided by a pattern is done very commonly there.
As far as I understood the proposed `DerefPure` would be used in that case behind the curtains of default binding modes, am I right about this? If not I would really like to know how this use case is covered.

**Comment by glaebhoerl** [comment]
There haven't been any developments on this front that I know of since then, if that's what you're thinking of.

**Comment by glaebhoerl** [comment]
There haven't been any developments on this front that I know of since then, if that's what you're thinking of.

**Comment by ice1000** [comment]
> @ice1000 There haven't been any developments on this front that I know of since then, if that's what you're thinking of.

Sad to hear that. Thanks for your information!

**Comment by jonhoo** [comment]
Another gentle bump to this issue — what's the current path towards stabilization? As far as I'm aware, this is still the only way to destructure across a `Box`.

**Comment by Karrq** [comment]
Yet another bump. Currently there's no way to destructure a box inside an enum variant in 1 step except by using this feature. Is there really no path towards stabilization? 
Triage?

**Comment by SimonSapin** [comment]
I understand the frustration, but at this point I think it’s pretty clear that this feature is not gonna be stabilized without another round of design work. This not a case of a feature that’s basically done and we just need to press the button.

The last formal Language Team decision around this feature was [RFC 469](https://github.com/rust-lang/rfcs/pull/469) in 2014, which made the feature unstable before Rust 1.0. (Or perhaps this was before there was a language team.) So the path forward if there is one likely involves a high-level proposal for how this feature should change (or why it shouldn’t), finding a member of the language team willing to champion this, and submitting a new RFC.

Regarding changes, the unstabilization RFC discusses extending the feature to types other than `Box`. This may be more tricky than simply calling `deref`: there’s precedent for matching a pattern that refers to a `const` being “structural” (fields are matched recursively) rather than calling `PartialEq::eq`. The trait is required to be derived, because it would be weird if matching and `==` gave different results.

```rust
#[derive(Eq)]
pub struct Foo(u32);

impl PartialEq for Foo {
    fn eq(&self, other: &Self) -> bool {
        (self.0 - other.0).abs() <= 3
    }
}

pub fn foo(x: Foo) {
    const FOUR: Foo = Foo(4);
    if let FOUR = x {}
}
```

[Playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=bc40eb5aa928f2bd4753a84d3c25bbe8), errors:

```rust
error: to use a constant of type `Foo` in a pattern, `Foo` must be annotated with `#[derive(PartialEq, Eq)]`
  --> src/lib.rs:12:12
   |
12 |     if let FOUR = x {}
   |            ^^^^
```

Also consider that the language team has limited bandwidth for design work. How important is this particular feature compared to ongoing work? https://github.com/servo/servo/commit/11bd81e1157dde8b3131766bf8ebecbb134e54ce is the commit that removed all uses of box patterns from Servo. The code becomes slightly less nice, but it’s not too bad IMO. This is **not** a case of being forced to rework the entire approach to a problem because of language limitations. Personally I’d be happy to delay box patterns indefinitely if it means delaying slightly less any of the big items like specialization, const generics, or generic associated types.

**Comment by Karrq** [comment]
Thank you for shedding some light on the status of this feature!

**Comment by nirvdrum** [comment]
@SimonSapin Thanks for the info. I do think box patterns are quite important. I'm still relatively new to the language, but I've been working on a compiler on-and-off for the past 18 months. The most natural way to represent an AST in Rust is as a recursive enum, which means needing to box values. The IR I'm using happens to be a recursive enum as well. Multiple phases of the compiler need to pattern match on specific structures and without box patterns, this grossly complicates the code. Rather than seeing a nice succinct pattern, you see part of a match with a bunch of manual unboxing then the next part of the match and so on.

It's not that the problem isn't solvable without box patterns. It's that I'm not convinced doing it without box patterns is even worthwhile doing in Rust. And that might be fine. Rust doesn't need to be the solution to every problem domain, so maybe OCaml is the right answer here. But, I thought I'd share how important this feature would be to someone like me. For me, not having this feature means I really need to convince myself I want to use Rust because it really affects the legibility, and thus maintainability, of my code.

**Comment by rpjohnst** [comment]
One alternative to box patterns in your case might be to allocate nodes in an arena, enabling you to use references instead.

That gives you support for default binding modes, which are even more succinct than box patterns, and is likely to be a performance improvement as well.

(Leaving this here so others with the same problem can see it.)

**Comment by bbqsrc** [comment]
Has there been any movement in this space since the last round of discussion, in particular perhaps about auto-deref for pattern matching instead of stabilising the `box` keyword for pattern matching?

Been implementing an AST and tearing out my hair. :smile:

**Comment by ice1000** [comment]
I think people are looking at "default-binding mode" now.

**Comment by brendanzab** [comment]
@ice1000 are you referring to #42640?

**Comment by ice1000** [comment]
> @ice1000 are you referring to #42640?

Yes.

**Comment by tema3210** [comment]
Can this be used to do an "owning pointer", like `fn A<T:  !Sized>(box data: T)`? This could mean, that we recieving a boxed, unsized value, but working with it like it was usized local. Is it a good adition to DST?

**Comment by fxdave** [comment]
As I was reading this, I have to tell that, this is really not the way of doing this. 
For exampe, we add box keyword and what about the other types? Will we add rc keyword, and refcell keyword and others, and even custom_cell keyword? Of course not. I'm not proposing any solution, but I personally wouldn't want to see this feature in this keyword form. There must be a better way.

**Comment by clarfonthey** [comment]
From what I understand, the box keyword works for rc and other types. It's basically the semantics of a deref-move, whatever that would mean.

**Comment by ice1000** [comment]
> As I was reading this, I have to tell that, this is really not the way of doing this. 
> For exampe, we add box keyword and what about the other types? Will we add rc keyword, and refcell keyword and others, and even custom_cell keyword? Of course not. I'm not proposing any solution, but I personally wouldn't want to see this feature in this keyword form. There must be a better way.

I guess it's #42640

**Comment by Nadrieril** [comment]
> From what I understand, the box keyword works for rc and other types. It's basically the semantics of a deref-move, whatever that would mean.

That's not the case currently, but there is a project group that is working to make something like this happen: https://github.com/rust-lang/project-deref-patterns

**Comment by joshtriplett** [comment]
Marking Box patterns (in the literal sense of the `box` keyword) as perma-unstable. If there's forward progress here, it'd be under the auspices of "deref patterns".

**Comment by safinaskar** [comment]
@deontologician, @jonhoo, @jcotton42, @ice1000, @Robbepop, @nirvdrum, @bbqsrc and everybody else interested!

I made proc macro library, which implements deref patterns right now, in stable Rust! 🎉

https://crates.io/crates/match_deref

It will work for any type, which implements Deref, for example Box, Rc and String.

My library is not drop-in replacement for box patterns, because I call Deref::deref internally, and Deref::deref takes a reference and returns a reference. I. e. with my lib you can do this: `match &Box::new(...) { Deref @ _ => ... }`, but you cannot do this: `match Box::new(...) { Deref @ _ => ... }`. I. e. moving a value out of a box is not supported.

Also currently there is no support for DerefMut and RefCell, but they can be trivially added. It is possible that I will add support for them (check periodically my lib on crates.io ). But you can simply download the code and just edit it in few places to get DerefMut and RefCell support

**Comment by KisaragiEffective** [comment]
maybe create new gate `min_box_pattern` and think about only Box on there? Is there blocker to stabilize?

**Comment by safinaskar** [comment]
@KisaragiEffective , there is already nearly-finished work for strings here: 
https://github.com/rust-lang/rust/pull/98914 . My opinion is this: we should create `DerefPure` as suggested in https://github.com/rust-lang/rust/pull/98914#issuecomment-1234545196 and then make String and Box both impls of DerefPure

**Comment by tgross35** [comment]
For anyone hunting for info - this https://github.com/rust-lang/rust/issues/87121 is the tracking issue for deref patterns, which eliminate the need for `box_patterns`. So, `box_patterns` will eventually be deprecated and not stabilized.

Support will probably be added one by one for std types (String, Vec, Box, Rc, Arc) until a trait is decided upon that enables use on custom types like `DerefPure` (which could only be those where `deref` has no side effects, [per zulip](https://rust-lang.zulipchat.com/#narrow/stream/213817-t-lang/topic/box.20pattern.20.26.20box.20syntax/near/322819832)). It seems like it will be a while before we're able to use the result though, since the changes look pretty wide reaching - see the preliminary implementation for String https://github.com/rust-lang/rust/pull/98914 that merged somewhat recently

**Comment by tgross35** [comment]
Feature has been removed in favor of deref patterns, see https://github.com/rust-lang/rust/pull/156749
